whetstone/compiler/src/main/kotlin/com/freeletics/mad/whetstone/parser/Reference.kt

```diff
@@ -75,3 +91,50 @@ internal val ClassReference.packageName: String
 private fun FqName.packageString(): String {
     return pathSegments().joinToString(separator = ".")
 }
+
+internal fun TypeName.asFunction1Parameter(): TypeName {
+    return Function1::class.asClassName().parameterizedBy(this, UNIT)
+}
+
+@OptIn(ExperimentalAnvilApi::class)
+internal fun ClassReference.resolveTypeParameter(
+    parameter: String,
+    superType: TypeReference,
+): TypeName {
+    // find the index of the type parameters that the current class has
+    // e.g. for a StateMachine and State this would return 0
+    val index = superType.asClassReference().typeParameters.indexOfFirst { it.name == parameter }
+    // this is the type that is used in the implementation
+    // e.g. for ... : StateMachine<S, A> this would be S
+    val unwrappedType = superType.unwrappedTypes[index]
+    // resolve the type using the implementation class
+    val resolved = unwrappedType.resolveGenericTypeOrNull(this)
+    if (resolved != null) {
+        return resolved.asTypeName()
+    }
```

Comment:
🤯

Reply:
This is already the simple version 😆 Initially I had `superTypeReference` return a list of type references for the whole hierarchy between the given class and the state machine interface and then this method would walk through the hierarchy and keep track of how the name of the type parameters change from one hierarchy level to another

Reply:
Nice! Well done! :shipit: